### PR TITLE
update-profile request

### LINKED ISSUE #2544: Support updating the public profile
Users have a public profile (sometimes registered via the zonefile) that contains public information like where users store their data for certain apps. Currently, the profile contains a schema definition for `Person`. However, there is no way for the user to add or change the personal details.

This feature request is a first step to support updating the public profile. It is limited to properties of the Person schema. 

Application can request to update the users profile, then the user has to confirm the change via UI like this:

Screenshot from 2022-07-07 13-19-43

This feature requires:
* handling a new type of request (update profile) from apps
* adding UI flow to confirm profile updates
* re-useing existing functions to upload the updated and signed profile

In connect (see https://github.com/hirosystems/connect/issues/246):
* adding new type of request (doUpdateProfile)

//cc @markmhx @yknl

### PR DESCRIPTION
**Pull request by friedger**
This PR
- adds support for update profile requests. These requests are sent from any app and requests that the public profile is updated. The profile is shown to the user before the public profile of the user is updated.
- fixes #2544 

image

cc/ @kyranjamie @fbwoolf @beguene @He1DAr

### CODE REVIEW

File: src/app/pages/update-profile-request/components/update-action.tsx
Comment (friedger) [inline src/app/pages/update-profile-request/components/update-action.tsx]: DEFAULT_PROFILE, fetchAccountProfileUrl, fetchProfileFromUrl, signAndUploadProfile and getHubInfo should be exported by wallet-sdk, or a combine method usable in useUpdateProfileSoftwareWallet

**Comment by markmhendrickson** [comment]
@friedger any way you could provide a test app of sorts that helps us QA this with various inputs?

**Comment by markmhendrickson** [comment]
Also I see this is in "Draft" – do you prefer we wait on our review until it's out of that state?

**Comment by friedger** [comment]
I would like to get early feedback. This is in a state that works.

To get a working app we would need connect support, wouldn't we?

File: src/app/store/profiles/requests.hooks.ts
Comment (kyranjamie) [inline src/app/store/profiles/requests.hooks.ts]: We're trying to move away from saving data in atoms (at all, but especially if it's ephemeral and doesn't really need to be shared globally within the app).

File: src/app/store/profiles/requests.hooks.ts
Comment (kyranjamie) [inline src/app/store/profiles/requests.hooks.ts]: ~~This doesn't actually update anything, it just reads from the initial params (as it should), so needs to be renamed.~~

See that this is the feature name, rather than the action update things. Maybe calling the feature with a verb isn't a good idea? `ProfileRequestUpdater`?

File: src/app/pages/update-profile-request/wallet-sdk-clone/profiles.ts
Comment (kyranjamie) [inline src/app/pages/update-profile-request/wallet-sdk-clone/profiles.ts]: Recognise this is for test purposes, but of course, we don't want to commit a clone. Would you make these changes in the lib itself? 🤔

File: src/app/pages/update-profile-request/update-profile-request.tsx
Comment (kyranjamie) [inline src/app/pages/update-profile-request/update-profile-request.tsx]: as this is a new feature, we shouldn't need this updater logic

File: src/app/pages/update-profile-request/components/profile-disclaimer.tsx
Comment (kyranjamie) [inline src/app/pages/update-profile-request/components/profile-disclaimer.tsx]: Can we name this something more specific to the feature?

**Review by kyranjamie — COMMENTED** [review]
Sorry for the delay leaving feedback @friedger. Left some high level comments, I think there's some duplication/tidying we'd want, but haven't mentioned atm as it's in draft.

Maybe @markmhx can speak to where this comes on our priorities to get merged.

File: src/app/store/profiles/requests.hooks.ts
Comment (friedger) [reply src/app/store/profiles/requests.hooks.ts]: @kyranjamie What is the preferred solution? Where/how enter the requests the app?

File: src/app/pages/update-profile-request/components/profile-disclaimer.tsx
Comment (friedger) [reply src/app/pages/update-profile-request/components/profile-disclaimer.tsx]: Copied from message-signing-disclaimer. Something like `UpdateProfileDisclaimer`?

File: src/app/pages/update-profile-request/wallet-sdk-clone/profiles.ts
Comment (friedger) [reply src/app/pages/update-profile-request/wallet-sdk-clone/profiles.ts]: Definitely, that is the plan. How do you handle newer versions of the wallet-sdk?

**Review by friedger — COMMENTED** [review]
I hope to get this merged sooner than later :-)

File: src/app/pages/update-profile-request/components/profile-disclaimer.tsx
Comment (kyranjamie) [reply src/app/pages/update-profile-request/components/profile-disclaimer.tsx]: Would you mind making a generic `DisclaimerLayout` component in `src/app/components` that both this one and message signing disclaimer also use.

File: src/app/pages/update-profile-request/components/update-action.tsx
Comment (kyranjamie) [inline src/app/pages/update-profile-request/components/update-action.tsx]: Imo this component should be dumb, and just fire an `onUpdateProfie()` and `onCancel` event. Can this logic then live in the parent?

File: src/app/pages/update-profile-request/components/update-profile-error-msg.tsx
Comment (kyranjamie) [inline src/app/pages/update-profile-request/components/update-profile-error-msg.tsx:20]: This is copied right? Can we just use the other component?

File: src/app/store/profiles/requests.hooks.ts
Comment (kyranjamie) [reply src/app/store/profiles/requests.hooks.ts]: You can already read this token from `useUpdateProfileRequestSearchParams()` so why do we need to save it in state?

File: src/app/store/profiles/requests.hooks.ts
Comment (kyranjamie) [inline src/app/store/profiles/requests.hooks.ts]: same here, we can read directly from the search param hook, so we don't need to rely on the rendering behaviour or component tree to persist and save in an atom

**Comment by friedger** [comment]
Wallet-sdk PR is here: https://github.com/hirosystems/stacks.js/pull/1351

**Comment by markmhendrickson** [comment]
@friedger Want to update this PR based on the comments above and let us know when it's ready for another review?

**Comment by kyranjamie** [comment]
@friedger it'd be great if we could add integration tests for this flow. Happy to run you through how to write/run these.

**Comment by friedger** [comment]
@kyranjamie thank you. I'll contact you for a date.

@markmhx I try to get this updated as quickly as possible. 😁

File: src/app/pages/update-profile-request/components/update-profile-error-msg.tsx
Comment (friedger) [reply src/app/pages/update-profile-request/components/update-profile-error-msg.tsx:20]: It is copied from /src/app/pages/transaction-request/components/minimal-error-message.tsx

Not sure how to extract a component with the Suspense. Would we need Suspense here as well?

**Comment by markmhendrickson** [comment]
@friedger I see you've been active on this PR but it's not clear it's ready for re-review. When it is, please reassign @kyranjamie as a reviewer and we'll triage 🙏

**Comment by friedger** [comment]
@markmhx In order to get integration tests going we need support by connect: https://github.com/hirosystems/connect/pull/274

File: src/app/store/profiles/requests.hooks.ts
Comment (friedger) [reply src/app/store/profiles/requests.hooks.ts]: not using atom anymore

File: src/app/pages/update-profile-request/wallet-sdk-clone/profiles.ts
Comment (friedger) [reply src/app/pages/update-profile-request/wallet-sdk-clone/profiles.ts]: wallet-sdk has been updated

**Comment by friedger** [comment]
image

Profile updating via test app works.

Next:
* fix missing properties in connect: https://github.com/hirosystems/connect/pull/277
* add proper integration tests with the help of @kyranjamie 
* define supported profile properties with @markmhx (?)

**Comment by friedger** [comment]
I have updated the screenshot in the description with the current design.

This PR just needs a new release of connect (https://github.com/hirosystems/connect/pull/279) using stacks.js from https://github.com/hirosystems/stacks.js/pull/1381

File: src/app/pages/update-profile-request/components/update-action.tsx
Comment (friedger) [reply src/app/pages/update-profile-request/components/update-action.tsx]: wallet-sdk has been updated

File: src/app/store/profiles/requests.hooks.ts
Comment (friedger) [reply src/app/store/profiles/requests.hooks.ts]: naming has been updated

File: src/app/pages/update-profile-request/components/update-profile-error-msg.tsx
Comment (friedger) [reply src/app/pages/update-profile-request/components/update-profile-error-msg.tsx:20]: component has been extracted

**Comment by markmhendrickson** [comment]
I'm glad to see this coming together! 

Can we improve the data formatting a bit, per the same guidance provided here? https://github.com/hirosystems/stacks-wallet-web/pull/2405#issuecomment-1123780529

Also is the red circle supposed to be the profile photo? I wonder if we should display the image URL raw like other attributes for now, and save displaying it for a future enhancement where we look at customizing this screen in general for standardized attributes?

**Update:** I see a few other UI nits, so I'll provide a design mockup soon to cover them all as guidance

**Comment by friedger** [comment]
Profile image is shown in addition to the plain data. For the image, it is `image[0].contentUrl`

**Comment by markmhendrickson** [comment]
@friedger Here's a mockup with some general formatting and copy improvements. What do you think?

image

[Figma version here](https://www.figma.com/file/6OdV4L3B01Ma0HlsXOyLqd/Profile-updates?node-id=1%3A947)

**Comment by friedger** [comment]
@markmhx Looks good. How are array items separated from each other? Like two images.

**Comment by markmhendrickson** [comment]
I've updated the above mockup to include guidance for arrays of objects: 

image

This seems to correspond to how arrays in structured messages are formatted:

Screen Shot 2022-10-24 at 14 03 25

**Comment by friedger** [comment]
`sameAs` is also an array. Between two arrays, there is only a `,`, not `],[`

**Comment by markmhendrickson** [comment]
Oops, fixed?

image

**Comment by friedger** [comment]
I have replaced the screenshot in the PR description with the current state.

Known issues:
* The screenshot shows "requested by unknown app" due to an older version of connect. This is fixed in https://github.com/hirosystems/connect/pull/279/commits/9d3c7401fa9166af2cf364d68565bf0d80cb0ff4. 
* connect-react does not yet support updating profiles.

This PR is now ready for review!

File: src/app/common/profiles/requests.ts
Comment (janniks) [inline src/app/common/profiles/requests.ts]: asdf 👀

Comment (friedger) [reply src/app/common/profiles/requests.ts]: That is a special error used at another place as well  (https://github.com/hirosystems/stacks-wallet-web/blob/427e592bef3cdd82c4bb63362efe1ee8a5cc36f0/src/app/common/signature/requests.ts#L86) :-)

What would be the proper error? The logic should prevent this error anyway.

**Comment by markmhendrickson** [comment]
Nice work! We'll review.

UI-wise, mind just bumping the image down so it top-aligns with the top of the header?

image

image

**Comment by friedger** [comment]
@markmhx Voila!
image

**Comment by markmhendrickson** [comment]
@friedger Thanks! Mind resolving the lint error?

**Comment by friedger** [comment]
@markmhx commits have been fixed. The failing integration test is part of dev branch I think

File: src/app/pages/update-profile-request/components/page-top.tsx
Comment (fbwoolf) [inline src/app/pages/update-profile-request/components/page-top.tsx]: Can just use the `Flex` component here from the UI library?

File: src/app/pages/update-profile-request/components/profile-box.tsx
Comment (fbwoolf) [inline src/app/pages/update-profile-request/components/profile-box.tsx]: A little confused why there are so many `Stack` elements used here?

File: src/app/pages/update-profile-request/components/profile-box.tsx
Comment (fbwoolf) [inline src/app/pages/update-profile-request/components/profile-box.tsx]: Can you make use of our UI library elements here and throughout?

File: tests/page-objects/profile-updating.page.ts
Comment (fbwoolf) [inline tests/page-objects/profile-updating.page.ts]: Is this meant to be left?

**Comment by fbwoolf** [comment]
One of us might have to cherry pick these commits to fix the failing test. Nice work @friedger, thanks for your help on this. I just had some minor comments about utilizing UI library elements in some places, but I'm sure one of us can address them if needed.

File: src/app/pages/update-profile-request/components/page-top.tsx
Comment (friedger) [reply src/app/pages/update-profile-request/components/page-top.tsx]: :+1: 
changed to `Flex`

File: src/app/pages/update-profile-request/components/profile-box.tsx
Comment (friedger) [reply src/app/pages/update-profile-request/components/profile-box.tsx]: I copied from structured message signature flow.
I updated it now to use `Box` instead.

File: tests/page-objects/profile-updating.page.ts
Comment (friedger) [reply tests/page-objects/profile-updating.page.ts]: removed

File: src/app/pages/update-profile-request/components/profile-box.tsx
Comment (friedger) [reply src/app/pages/update-profile-request/components/profile-box.tsx]: Are there any guidelines? I am not sure what component would be appropriate here.

File: src/app/pages/update-profile-request/components/page-top.tsx
Comment (kyranjamie) [inline src/app/pages/update-profile-request/components/page-top.tsx]: Use `whenChainId` helper

File: src/app/pages/update-profile-request/components/profile-box.tsx
Comment (kyranjamie) [reply src/app/pages/update-profile-request/components/profile-box.tsx]: This would be `<Box textIndex="" ml="loose" />`

File: src/app/pages/update-profile-request/components/profile-box.tsx
Comment (kyranjamie) [inline src/app/pages/update-profile-request/components/profile-box.tsx]:
```suggestion
    <Box minHeight="260px">
      <Stack
        border="4px solid"
        paddingBottom={'8px'}
        borderColor={color('border')}
        borderRadius="20px"
        backgroundColor={color('border')}
      >
        <Box py="loose" px="loose" spacing="loose" borderRadius="16px" backgroundColor={'white'}>
          <Box spacing="base-tight">
            <Properties p={profile._profile} />
          </Box>
        </Box>
      </Stack>
    </Box>
```

File: package.json
Comment (edu-stx) [inline package.json]: What's the need for using a lower version of this package?

Comment (friedger) [reply package.json]: That is a version build on top of 7.0.0. It contains changes that have not yet been released as a proper version

File: src/app/pages/update-profile-request/components/profile-box.tsx
Comment (friedger) [reply src/app/pages/update-profile-request/components/profile-box.tsx]: @kyranjamie 

Before:
image

After:
image

File: src/app/pages/update-profile-request/components/profile-box.tsx
Comment (friedger) [reply src/app/pages/update-profile-request/components/profile-box.tsx]: applied

File: src/app/pages/update-profile-request/components/page-top.tsx
Comment (friedger) [reply src/app/pages/update-profile-request/components/page-top.tsx]: done

**Comment by markmhendrickson** [comment]
@kyranjamie Mind helping to generate a test build here for QA? Is it not getting generated due to the Snyk error?

**Comment by paradigma-cl** [comment]
Waiting for the approval of the update-profile request functionality as it has become essential for our Dapp CrossCheck xck.app next release. We have already integrated it into the dapp, and we are still conducting tests, which we hope to share on Monday.  The actual version stores all the profile data in the GAIA, this new functionality gives more flexibility to multi-app compatibility. Now happily feeling the pressure for around 10K new users for xck.app, so eager to advance in this topic, before distribution. Thanks!

**Comment by markmhendrickson** [comment]
Test build getting generated here: https://github.com/hirosystems/stacks-wallet-web/pull/2829

**Comment by friedger** [comment]
@markmhx https://github.com/hirosystems/stacks-wallet-web/pull/2829 is not using the latest code that includes the fixes for the import organisation of the dev branch

**Comment by friedger** [comment]
How do we deal with known issues of the current code? @markmhx 
* app name not shown
* not possible to delete properties of profile

**Review by fbwoolf — APPROVED** [review]

**Comment by paradigma-cl** [comment]
Great!  Thank you very much

**Comment by markmhendrickson** [comment]
> https://github.com/hirosystems/stacks-wallet-web/pull/2829 is not using the latest code that includes the fixes for the import organisation of the dev branch

cc @kyranjamie who can help 

> How do we deal with known issues of the current code? (app name not shown, not possible to delete properties of profile)

Per showing the app name, want to create an issue in the connect repo? That seems to be what's missing here. 

Will this wallet-side enhancement work without further changes once connect supports the passing of app names on its end? If so, I feel comfortable releasing this wallet enhancement first even if in the meantime it'll say "Request by an unknown app"

Per deleting properties, can an app use profile updates to effectively delete properties by submitting empty values? If so, I feel comfortable releasing this enhancement as-is. But if not, we should figure out how to handle deletions before releasing.

**Comment by kyranjamie** [comment]
Updated branch

**Comment by paradigma-cl** [comment]
> > #2829 is not using the latest code that includes the fixes for the import organisation of the dev branch
> 
> cc @kyranjamie who can help
> 
> > How do we deal with known issues of the current code? (app name not shown, not possible to delete properties of profile)
> 
> Per showing the app name, want to create an issue in the connect repo? That seems to be what's missing here.

Probably it is a good idea to create the issue in the connect repo.


> Will this wallet-side enhancement work without further changes once connect supports the passing of app names on its end? If so, I feel comfortable releasing this wallet enhancement first even if in the meantime it'll say "Request by an unknown app"

In the meantime, I believe it could be alright to have this message 

> Per deleting properties, can an app use profile updates to effectively delete properties by submitting empty values? If so, I feel comfortable releasing this enhancement as-is. But if not, we should figure out how to handle deletions before releasing.

Probably, it could be a solution to have empty values in the meantime.  But as the schemas could be sometimes long, some properties may get deprecated, so having a way to delete those empty or not-used properties could be useful. 

Thank you!

**Comment by friedger** [comment]
> Per showing the app name, want to create an issue in the connect repo? That seems to be what's missing here.

This has already been fixed in connect.

@kyranjamie where is the connect fork for this that includes doProfileUpdate? https://github.com/hirosystems/connect/pull/279

@markmhx Providing an undefined property to delete a value, works already. For deleting the property itself, we might want to add a new mechanism in the future, also some solutions for deep merge. As a MVP, the main focus is on adding and updating properties, IMHO.

**Comment by markmhendrickson** [comment]
cc @kyranjamie @janniks re: Friedger's connect fork question above ☝️ 

Sounds good then if deletions are effectively possible with this release.

@kyranjamie mind kicking https://github.com/hirosystems/stacks-wallet-web/pull/2829? The test build isn't getting generated there for some reason.

**Comment by kyranjamie** [comment]
The build is up to date, the problem is the `yarn.lock` is out of date with the package json. See errors logs for all the Code check jobs in this PR. @friedger can you fix this and ping me to update the other build?

**Comment by friedger** [comment]
Commit 146d867b8895776cd4f7296c2fee4a2a4d7e2ba6 removed the alpha versions for connect and stacks.js. Should I fix that?

**Comment by kyranjamie** [comment]
> Commit [146d867](https://github.com/hirosystems/stacks-wallet-web/commit/146d867b8895776cd4f7296c2fee4a2a4d7e2ba6) removed the alpha versions for connect and stacks.js. Should I fix that?

it's just a packaging issue. Your yarn.lock is out of sync with your package json. Run `yarn` and commit the resulting changes to the `yarn.lock`.

**Comment by friedger** [comment]
@kyranjamie Tests failed, build fails due to problems with dependencies.

**Comment by markmhendrickson** [comment]
@friedger anything you need from us to help resolve the dependency problems?

**Comment by friedger** [comment]
@markmhx All good for QA now!

**Comment by markmhendrickson** [comment]
@friedger it looks like we have one last lint check to resolve?

**Comment by friedger** [comment]
@markmhx What does the lint rule mean? What needs to be changed? Is this something related to the project setup? I don't see how I can fix this.

`error no-non-package-json: src/shared/actions/finalize-profile-update.ts → node_modules/@stacks/profile/dist/esm/index.js`

**Comment by kyranjamie** [comment]
This error means you're using `@stacks/profile` without it being listed in the package JSON

**Comment by kyranjamie** [comment]
Closing in favour of https://github.com/hirosystems/stacks-wallet-web/pull/2829

**Pull request closed without merging**
